datasource/src/main/scala/quasar/physical/mongo/decoder.scala

```diff
@@ -0,0 +1,217 @@
+      case BsonType.DOUBLE => QDouble
+      case BsonType.DECIMAL128 => {
+        val dec128: Decimal128 = bson.asDecimal128().getValue()
+        if (dec128.isNaN()) {
+          QNull
+        } else if (dec128.isInfinite()) {
+          QNull
+        } else try {
+          val decimal: JDecimal = dec128.bigDecimalValue()
+          if (decimal.scale() < 1 && decimal.compareTo(moreThanMaxLong) < 1 && decimal.compareTo(lessThanMinLong) > -1) {
+            QLong
+          } else {
+            if (decimal.abs().compareTo(tinyDbl) < 0) {
+              QReal
+            } else
+              if (decimal.compareTo(moreThanMaxDbl) < 1 && decimal.compareTo(lessThanMinDbl) > -1) {
+                QDouble
+              } else {
+                QReal
+              }
+          }
+        } catch {
+          case e: ArithmeticException => QLong
+        }
+      }
+      case BsonType.BOOLEAN => QBoolean
+      case BsonType.OBJECT_ID => QMeta
+      case BsonType.DB_POINTER => QMeta
+      case BsonType.BINARY => QMeta
+      case BsonType.DATE_TIME => QOffsetDateTime
+      case BsonType.SYMBOL => QMeta
+      case BsonType.REGULAR_EXPRESSION => QMeta
+      case BsonType.JAVASCRIPT => QMeta
+      case BsonType.NULL => QNull
+      case BsonType.TIMESTAMP => QMeta
+      case BsonType.UNDEFINED => QNull
+      case BsonType.JAVASCRIPT_WITH_SCOPE => QMeta
+      case BsonType.MAX_KEY => QMeta
+      case BsonType.MIN_KEY => QMeta
+      case BsonType.END_OF_DOCUMENT => QNull
+    }
+
+    type ArrayCursor = Iterator[BsonValue]
+
+    override def getArrayCursor(bson: BsonValue): ArrayCursor = bson match {
+      case arr: BsonArray => arr.getValues().iterator()
+    }
+    override def getArrayAt(cursor: ArrayCursor): BsonValue =
+      cursor.next()
+    override def hasNextArray(cursor: ArrayCursor): Boolean =
+      cursor.hasNext()
+    override def stepArray(cursor: ArrayCursor): ArrayCursor =
+      cursor
+
+    type ObjectCursor = BsonDocumentCursor
+
+    override def getObjectCursor(bson: BsonValue): ObjectCursor = bson match {
+      case obj: BsonDocument => new BsonDocumentCursor(obj)
+    }
+    override def getObjectKeyAt(cursor: ObjectCursor): String =
+      cursor.getObjectKeyAt()
+    override def getObjectValueAt(cursor: ObjectCursor): BsonValue =
+      cursor.getObjectValueAt()
+    override def hasNextObject(cursor: ObjectCursor): Boolean =
+      cursor.hasNext()
+    override def stepObject(cursor: ObjectCursor): ObjectCursor =
+      cursor.stepObject()
+
+    override def getBoolean(bson: BsonValue): Boolean = bson match {
+      case bool: BsonBoolean => bool.getValue()
+    }
+    override def getDouble(bson: BsonValue): Double = bson match {
+      case num: BsonNumber => num.doubleValue()
+    }
+    @SuppressWarnings(Array("org.wartremover.warts.Equals"))
+    override def getLong(bson: BsonValue): Long = bson match {
+      case bsonDecimal: BsonDecimal128 =>
+        try {
```

Comment:
Because of `NEGATIVE_ZERO` there is no way to know if it's one of `-0`s from `Decimal128` and it throws `ArithmeticException` during conversion to `BigDecimal`